<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Counts records in data from the Alliance of Genome Resources. Counts can
be ascribed to the species the record is associated with or the Model
Organism Database (MOD) that curated it, optionally by object type.
count_alliance_records() was primarily designed to count records in the
Alliance Disease Associations File. There is no guarantee that any/all
options will work for other files."><title>Count Alliance Records — count_alliance_records • DO.utils</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Count Alliance Records — count_alliance_records"><meta property="og:description" content="Counts records in data from the Alliance of Genome Resources. Counts can
be ascribed to the species the record is associated with or the Model
Organism Database (MOD) that curated it, optionally by object type.
count_alliance_records() was primarily designed to count records in the
Alliance Disease Associations File. There is no guarantee that any/all
options will work for other files."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DO.utils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.9.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/obtain_use_records.html">Assessing Resource Use: Obtaining Use Records</a>
    <a class="dropdown-item" href="../articles/use-reticulate-python.html">Using DO.utils with reticulate and Python</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Count Alliance Records</h1>
      
      <div class="d-none name"><code>count_alliance_records.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Counts records in data from the Alliance of Genome Resources. Counts can
be ascribed to the species the record is associated with or the Model
Organism Database (MOD) that curated it, optionally by object type.
<code>count_alliance_records()</code> was primarily designed to count records in the
Alliance Disease Associations File. <em>There is no guarantee that any/all</em>
<em>options will work for other files.</em></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">count_alliance_records</span><span class="op">(</span></span>
<span>  <span class="va">alliance_tbl</span>,</span>
<span>  term_subset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  by_type <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pivot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  record_lvl <span class="op">=</span> <span class="st">"disease-object"</span>,</span>
<span>  assign_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"species"</span>, <span class="st">"curator"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>alliance_tbl</dt>
<dd><p>a dataframe derived from Alliance data (usually a
<a href="https://www.alliancegenome.org/downloads" class="external-link">downloaded .tsv file</a>)</p></dd>


<dt>term_subset</dt>
<dd><p>character vector of DOIDs to limit counts to</p></dd>


<dt>by_type</dt>
<dd><p>logical indicating whether to count by Alliance object type
(i.e. gene, allele, model)</p></dd>


<dt>pivot</dt>
<dd><p>logical indicating whether to pivot values to type columns;
ignored if by_type = FALSE.</p></dd>


<dt>record_lvl</dt>
<dd><p>a string indicating the desired specificity of records.</p></dd>


<dt>assign_to</dt>
<dd><p>how to assign records when counting; one of "species" or
"curator" (i.e. the organization responsible for curating the record)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A summary tibble with the count of unique object annotations defined by
<code>record_lvl</code>, aggregated according to species/curator (<code>assign_to</code>) and,
optionally, object type (<code>by_type</code>).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The type of record information to use in counting should be specified with
<code>record_lvl</code> which accepts the following values:</p><ul><li><p>"full_record" counts full non-duplicate records</p></li>
<li><p>"disease-object" counts unique disease-object combinations</p></li>
<li><p>"disease" counts unique diseases</p></li>
<li><p>"object" counts unique MOD objects (i.e. gene, allele, model identifiers)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="note">NOTE<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    

<p>For disease-related data, some exact duplicates (reason unknown) and records
that differ by seemingly unimportant information (e.g. only the date differs)
have existed. These types of duplicates are removed prior to record counts.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by J. Allen Baron.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

